### PR TITLE
Cloudtrust 1617 update account

### PR DESCRIPTION
**Pull request by harture**
Needs an update of keycloka-client dependency before merge

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -31,7 +31,8 @@
 
 [[constraint]]
   name = "github.com/cloudtrust/keycloak-client"
-  version = "=v1.0-rc3"
+  #version = "=v1.0-rc3"
+  branch="CLOUDTRUST-1617_updateAccount"
```
Comment (fperot74) [inline Gopkg.toml]: Reminder: to be updated before merging this PR (and remove equal sign in version definition)

File: api/account/api.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	userRep.FirstName = user.FirstName
+	userRep.LastName = user.LastName
+
+	var attributes = make(map[string][]string)
+
+	if user.PhoneNumber != nil {
+		attributes["phoneNumber"] = []string{*user.PhoneNumber}
+	}
+
+	if len(attributes) > 0 {
```
Comment (fperot74) [inline api/account/api.go]: Fine if you plan to add attributes... else, this can be moved in the if block above

File: api/account/api.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	return userRep
+}
+
+// Validators
+
+// Validate is a validator for AccountRepresentation
+func (user AccountRepresentation) Validate() error {
+	if user.Username != nil && !matchesRegExp(*user.Username, RegExpUsername) {
+		return errors.New("Invalid username")
```
Comment (fperot74) [inline api/account/api.go]: To be updated if this PR is merged after cloudtrust/keycloak-bridge#114

File: pkg/account/component.go [hunk 4]
```diff
@@ -68,3 +78,95 @@ func (c *component) UpdatePassword(ctx context.Context, currentPassword, newPass
...
+	oldUserKc, err := c.keycloakClient.GetAccount(accessToken, realm)
+	if err != nil {
+		c.logger.Warn("err", err.Error())
+		return err
+	}
+
+	var emailVerified, phoneNumberVerified *bool
+
+	// when the email changes, set the EmailVerified to false
+	if user.Email != nil && oldUserKc.Email != nil && *oldUserKc.Email != *user.Email {
```
Comment (fperot74) [inline pkg/account/component.go]: Copied/pasted from management/component.go ? Couldn't it be factorized ?

File: pkg/account/component_test.go [hunk 5]
```diff
@@ -88,3 +96,254 @@ func TestUpdatePasswordWrongPwd(t *testing.T) {
...
+	var email = "toto@elca.ch"
+	var enabled = true
+	var emailVerified = true
+	var firstName = "Titi"
+	var lastName = "Tutu"
+	var phoneNumber = "+41789456"
+	var phoneNumberVerified = true
+	var label = "Label"
+	var gender = "M"
+	var birthDate = "01/01/1988"
```
Comment (fperot74) [inline pkg/account/component_test.go]: Why 1988 ? 😄

Comment (bsoniam) [reply pkg/account/component_test.go]: People are not trustworthy, they have the tendency to leak information related to them :D

File: api/account/swagger-api_account.yaml [hunk 6]
```diff
@@ -26,6 +26,30 @@ paths:
...
+        200:
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/User'
+    post:
+      tags:
+      - Account
+      summary: Update account representation of the current user.
```
Comment (bsoniam) [inline api/account/swagger-api_account.yaml]: no dot

File: api/account/api.go (hunk 3 above)
Comment (bsoniam) [reply api/account/api.go]: I believe it will be merged before....

Comment (harture) [reply api/account/api.go]: We will have to merge this one before

File: pkg/account/component.go (hunk 4 above)
Comment (harture) [reply pkg/account/component.go]: Not exactly a copy paste. It diverged as it use an AccountRepresentation which has no attributes. So I don't see a clean way to factorize it.

File: pkg/account/component_test.go (hunk 5 above)
Comment (harture) [reply pkg/account/component_test.go]: Just a wonderful year ;)

**Pull request merged**